letta/agent.py

```diff
@@ -1343,6 +1344,10 @@ def save_agent(agent: Agent, ms: MetadataStore):
     else:
         ms.create_agent(agent_state)
 
+    for tool in agent.tools:
+        if ms.get_tool(tool_id=tool.id) is None:
```

Comment:
You should actually get the tool by name like `self.ms.get_tool(tool_name=tool_name, user_id=user_id)` where I think you can get `user_id` from `agent_state.user_id`. 

This is because we don't allow duplicate names for tools, inside a user scope.

Reply:
Fixed!